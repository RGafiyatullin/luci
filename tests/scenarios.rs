use insta::{assert_debug_snapshot, assert_yaml_snapshot};
use luci::{
    execution_graph::ExecutionGraph,
    messages::{Messages, Mock},
    scenario::Scenario,
};
use test_case::test_case;

#[test_case("01", true)]
#[test_case("02", true)]
#[test_case("03", true)]
#[test_case("04", false)]
fn run(name: &str, build_graph: bool) {
    let input = format!("tests/scenarios/{name}.yaml");
    let scenario: Scenario =
        serde_yaml::from_str(std::fs::read_to_string(input).expect("read file").as_str())
            .expect("serde::de");
    assert_debug_snapshot!(format!("{name}.debug"), scenario);
    assert_yaml_snapshot!(format!("{name}.yaml"), scenario);

    let messages = Messages::new()
        .with(Mock::msg("one::two::Three"))
        .with(Mock::req("one::two::IsAThree"))
        .with(Mock::msg("protocol_basic::Start"))
        .with(Mock::msg("protocol_basic::Started"))
        .with(Mock::msg("protocol_basic::KeepAlive"));

    if build_graph {
        let _exec_graph = ExecutionGraph::builder(messages)
            .build(&scenario)
            .expect("build exec-graph");
    }
}
