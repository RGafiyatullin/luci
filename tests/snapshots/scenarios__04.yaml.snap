---
source: tests/scenarios.rs
expression: scenario
---
types:
  - use: "protocol_basic::KeepAlive"
    as: KeepAlive
sub:
  - load: 04/startup.yaml
    as: startup
cast:
  - a
  - b
events:
  - id: total-test-duration
    delay:
      for: 15s
      step: 1s
  - id: standard-startup-routine
    call:
      sub: startup
      cast:
        - use: a
          as: subject
        - use: b
          as: object
      input:
        dst: $component_id
        src:
          literal: the-thingy
      output:
        dst: $the_thingys_launch_id
        src:
          bind: $launch_id
  - id: the-thingy-is-alive
    require: reached
    recv:
      type: KeepAlive
      data:
        bind:
          launch_id: $the_thingys_launch_id
      from: b
      to: a
